simplesat/constraints/constraint_modifiers.py

```diff
@@ -0,0 +1,151 @@
+    # This just eats the 'version' argument
+    return Any()
+
+
+def LEQ_LEAST_UPPER_BOUND(version):
+    new_version = EnpkgVersion.from_upstream_and_build(
+        str(version.upstream), MAX_BUILD)
+    return LEQ(new_version)
+
+
+ALLOW_NEWER_MAP = {
+    Any: Any_,
+    Equal: GEQ,
+    Not: Not,
+    GEQ: GEQ,
+    GT: GT,
+    LEQ: Any_,
+    LT: Any_,
+    EnpkgUpstreamMatch: GEQ,
+}
+
+ALLOW_OLDER_MAP = {
+    Any: Any_,
+    Equal: LEQ,
+    Not: Not,
+    GEQ: Any_,
+    GT: Any_,
+    LEQ: LEQ,
+    LT: LT,
+    EnpkgUpstreamMatch: LEQ_LEAST_UPPER_BOUND,
+}
+
+ALLOW_ANY_MAP = {
+    Any: Any_,
+    Equal: Any_,
+    Not: Not,
+    GEQ: Any_,
+    GT: Any_,
+    LEQ: Any_,
+    LT: Any_,
+    EnpkgUpstreamMatch: Any_,
+}
+
+
+def iterable_to_set(container):
+    """ Return a set from an iterable, being careful not to disassemble
+    strings.
+        >>> iterable_to_set(['foo'])
+        set(['foo'])
+        >>> iterable_to_set('foo')
+        set(['foo'])
+    """
+    if isinstance(container, six.string_types):
+        container = (container,)
+    return set(container)
+
+
+_defaults = dict(default=(), convert=iterable_to_set,
+                 validator=instance_of(set))
+
+
+@attributes
+class ConstraintModifiers(object):
+    allow_newer = attr(**_defaults)
+    allow_any = attr(**_defaults)
+    allow_older = attr(**_defaults)
+
+    def asdict(self):
+        return asdict(self)
+
+    @property
+    def targets(self):
+        return set.union(self.allow_newer, self.allow_any, self.allow_older)
+
+
+def _transform_install_requirement(
+        requirement, allow_newer=None, allow_any=None, allow_older=None):
+    """If any of the modifier rules apply, return a new Requirement with
+    modified constraints, otherwise return the original requirement.
+    """
+
+    name = requirement.name
+    original_constraints = constraints = requirement._constraints._constraints
+    transformers = (
+        (allow_older or (), ALLOW_OLDER_MAP),
+        (allow_newer or (), ALLOW_NEWER_MAP),
+        (allow_any or (), ALLOW_ANY_MAP),
+    )
+
+    modified = False
+    for names, type_map in transformers:
+        if name in names:
+            modified = True
+            constraints = _transform_constraints(constraints, type_map)
+
+    if modified and constraints != original_constraints:
+        # Remove duplicate constraints
+        constraints = tuple(OrderedDict.fromkeys(constraints).keys())
+        return InstallRequirement(name, constraints)
+
+    return requirement
+
+
+def _transform_constraints(constraints, type_map):
+    return tuple(
+        type_map[type(c)](getattr(c, 'version', None))
+        for c in constraints
+    )
+
+
+class _TransformRequirement(object):
+
+    @staticmethod
+    def __call__(requirement, **kw):
+        if isinstance(requirement, InstallRequirement):
+            return _transform_install_requirement(requirement, **kw)
+        elif requirement.has_any_version_constraint:
```

Comment:
This will allow non-`InstallRequirement` objects to pass through, as long as the transformation is a no-op.

Should it be a `TypeError` instead?